### PR TITLE
Allow ndarrays in `IndexFilter.indices` and add "cross" tests

### LINKED ISSUE #13134: [BUG] IndexFilter indices no longer accepts numpy.ndarray
### Software versions

Python version      :  3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:35:26) [GCC 10.4.0]
IPython version     :  8.5.0
Tornado version     :  6.2
Bokeh version       :  3.1.1
BokehJS static path :  /home/zpincus/miniconda3/envs/hxenv/lib/python3.10/site-packages/bokeh/server/static
node.js version     :  (not installed)
npm version         :  (not installed)
Operating system    :  Linux-4.19.0-23-cloud-amd64-x86_64-with-glibc2.28

### Browser name and version

_No response_

### Jupyter notebook / Jupyter Lab version

_No response_

### Expected behavior

This is basically a dupe of #12624, except for `IndexFilter.indices` instead of a data source's `selected.indices`

Before bokeh 3, both of the last two lines worked:
```python
    source = models.ColumnDataSource(df)
    source.selected.indices = np.flatnonzero(df.value == select_value)
    filter = models.IndexFilter(np.flatnonzero(df.value == filter_value)
```
After #12624 was fixed, 
```python
    source.selected.indices = np.flatnonzero(df.value == select_value)
```
works, but
```python
    filter = models.IndexFilter(np.flatnonzero(df.value == filter_value)
```
still does not. 

The expected behavior is returned with
```python
    filter = models.IndexFilter(list(np.flatnonzero(df.value == filter_value))
```

### Observed behavior

no plot is shown and a js error appears on the console. 

### Example code

```Python
import numpy as np
from bokeh.layouts import gridplot
from bokeh.models import CDSView, ColumnDataSource, IndexFilter
from bokeh.plotting import figure, show

source = ColumnDataSource(data=dict(x=[1, 2, 3, 4, 5], y=[1, 2, 3, 4, 5]))
view = CDSView(filter=IndexFilter(np.array([0, 2, 4])))
p1 = figure(height=300, width=300)
p1.circle(x="x", y="y", size=10, hover_color="red", source=source)

p2 = figure(height=300, width=300)
p2.circle(x="x", y="y", size=10, hover_color="red", source=source, view=view)

show(gridplot([[p1, p2]]))
```

### Stack traceback or browser console output

_No response_

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by mattpap**
This is a trivial bug fix. However, it will take the opportunity here to add a new kind of integration tests, which simultaneously will allow to test JSON output from bokeh and bokehjs' interpretation of that output, allowing for better integration testing between bokeh and bokehjs. This will not use selenium or any kind of JS testing on the Python side.

fixes #13134

### CODE REVIEW

**Comment by bryevdv** [comment]
@mattpap will you be adding docs for these new tests in this PR? All questions I currently have: what are they, what do they do, what are their exact benefits, how do they work, how are they created/configured, how can they be run locally...

**Comment by mattpap** [comment]
Cross integration tests are similar to bokehjs' integration tests. However, instead of them being written in TypeScript and run immediately, they are written in Python, run, generate JSON output and then run in bokehjs using `Document.from_json()`. The serialized output is stored in the repository under `tests/baselines/cross`, similarly to any other textual baselines we have (defaults, `*.blf` files), i.e. tests are run (`pytest tests/test_cross.py`), baselines generated and committed, and tests are re-run to verify.

This will allow us to integration test bokeh and bokehjs without selenium (or similar), while reusing bokehjs' integration test infrastructure. Storing generated JSON will allow us, for the first time, to integration test the protocol. The expectation is that those tests will be created with the intent to integration test bokeh and bokehjs, not just bokehjs' UI, etc., and the protocol. 

Image diff can be skipped for those tests with `it.no_image(...)` on bokehjs' side. Test cases, despite them being written in Python, need to follow design guidelines for bokehjs' integration tests, which means they have to be carefully designed to not use any external resources (those can be provided by bokehjs' devtools server), be fully deterministic, etc. Random data is permitted, because built-in and numpy's random generators are seeded. Additionally they should be written using models (no plotting API).

**Comment by mattpap** [comment]
This is ready for review. ~~Support for Windows needs to be still completed.~~ I skipped testing on Windows. The code is equivalent to `test_examples`, which I don't know if it works on Windows or not, but I don't have currently any ability to test on Windows.

File: docs/bokeh/source/docs/dev_guide/testing.rst [hunk 1]
```diff
@@ -192,6 +192,33 @@ Integration tests
...
+    This is a variant of integration tests where on Bokeh's side a Python
+    code sample (a test case) is run, which produces JSON output with the
+    serialized document. That JSON is then stored in the repository under
+    ``tests/baselines/cross``. When adding a new test case, run:
+
+    .. code-block:: sh
+
+        pytest tests/test_cross.py
+
+    and then commit any new baselines and re-run tests. Only commited
```
Comment (ianthomas23) [inline docs/bokeh/source/docs/dev_guide/testing.rst]:
```suggestion
    and then commit any new baselines and re-run tests. Only committed
```

**Review by ianthomas23 — APPROVED** [review]
I am approving this to unblock the queue of PRs for the 3.2 release. There is just one typo.

Documentation has been added to the dev guide which I think addresses @bryevdv's comment. I'd like to try out the process of adding new cross tests but that can wait for the moment; if it indicates any problems with the process or documentation we can reconsider changes then.

For full benefit I assume we would want to test on Windows, which probably needs someone to volunteer...

**Comment by bryevdv** [comment]
> I think addresses @bryevdv's comment

Yes thank you @mattpap for the docs and explanation. Quick bikeshed: is there a better name? e.g. "cross runtime" or "playback" tests? I do feel simply "cross" is a bit too generic (cross-*what*)

**Comment by mattpap** [comment]
> Is there a better name? e.g. "cross runtime" or "playback" tests? I do feel simply "cross" is a bit too generic (cross-what)

I would just call them integration tests (or a subset of integration tests), especially given that in bokehjs they are a subset of its integration tests. It would rename the existing bokeh's integration tests to maybe legacy integration tests. Most of them test bokehjs' features and need to be migrated to bokehjs' integration tests. Those actually testing bokeh would be converted into the currently named cross integration tests. The remainder would be server integration tests (tests that actually test the server).

**Comment by bryevdv** [comment]
"integration" also seems like a fine category tho then I think it's still worth being explicit as you seem to allude to (python standalone vs server, etc)

**Comment by ianthomas23** [comment]
I've had a quick look on Windows and a simple change allows two of the three tests to pass. The `stdout` return from `_run_test_case` has Windows line endings `"\r\n"` which if converted to `"\n"` allows the first and last tests to pass.

The middle test is more problematic, I'm getting a timeout on the `subprocess` call that runs the test. I assume the timeout isn't the actual problem. Here's the full log
```
(bkdev) PS C:\Users\iant\github\bokeh> pytest -v .\tests\test_cross.py
======================================================= test session starts ========================================================
platform win32 -- Python 3.10.11, pytest-7.3.1, pluggy-1.0.0 -- C:\Users\iant\miniconda3\envs\bkdev\python.exe
cachedir: .pytest_cache
metadata: {'Python': '3.10.11', 'Platform': 'Windows-10-10.0.22621-SP0', 'Packages': {'pytest': '7.3.1', 'pluggy': '1.0.0'}, 'Plugins': {'anyio': '3.6.2', 'asyncio': '0.21.0', 'cov': '4.0.0', 'html': '3.2.0', 'metadata': '2.0.4', 'timeout': '2.1.0', 'xdist': '3.3.1'}}
rootdir: C:\Users\iant\github\bokeh
configfile: pyproject.toml
plugins: anyio-3.6.2, asyncio-0.21.0, cov-4.0.0, html-3.2.0, metadata-2.0.4, timeout-2.1.0, xdist-3.3.1
asyncio: mode=strict
timeout: 60.0s
timeout method: thread
timeout func_only: False
collected 3 items

tests/test_cross.py::test_cross[regressions\\issue_11694.py-config0] PASSED                                                   [ 33%]
tests/test_cross.py::test_cross[regressions\\issue_11930.py-config1]
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ Timeout ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Stack of MainThread (9956) ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  File "C:\Users\iant\miniconda3\envs\bkdev\Scripts\pytest-script.py", line 9, in <module>
    sys.exit(console_main())
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\_pytest\config\__init__.py", line 189, in console_main
    code = main()
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\_pytest\config\__init__.py", line 166, in main
    ret: Union[ExitCode, int] = config.hook.pytest_cmdline_main(
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\pluggy\_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\pluggy\_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\pluggy\_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\_pytest\main.py", line 316, in pytest_cmdline_main
    return wrap_session(config, _main)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\_pytest\main.py", line 269, in wrap_session
    session.exitstatus = doit(config, session) or 0
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\_pytest\main.py", line 323, in _main
    config.hook.pytest_runtestloop(session=session)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\pluggy\_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\pluggy\_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\pluggy\_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\_pytest\main.py", line 348, in pytest_runtestloop
    item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\pluggy\_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\pluggy\_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\pluggy\_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\_pytest\runner.py", line 114, in pytest_runtest_protocol
    runtestprotocol(item, nextitem=nextitem)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\_pytest\runner.py", line 133, in runtestprotocol
    reports.append(call_and_report(item, "call", log))
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\_pytest\runner.py", line 222, in call_and_report
    call = call_runtest_hook(item, when, **kwds)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\_pytest\runner.py", line 261, in call_runtest_hook
    return CallInfo.from_call(
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\_pytest\runner.py", line 341, in from_call
    result: Optional[TResult] = func()
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\_pytest\runner.py", line 262, in <lambda>
    lambda: ihook(item=item, **kwds), when=when, reraise=reraise
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\pluggy\_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\pluggy\_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\pluggy\_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\_pytest\runner.py", line 169, in pytest_runtest_call
    item.runtest()
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\_pytest\python.py", line 1799, in runtest
    self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\pluggy\_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\pluggy\_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\pluggy\_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\site-packages\_pytest\python.py", line 194, in pytest_pyfunc_call
    result = testfunction(**testargs)
  File "C:\Users\iant\github\bokeh\tests\test_cross.py", line 69, in test_cross
    status, _, out, err = _run_test_case(CASE_DIR / case_path)
  File "C:\Users\iant\github\bokeh\tests\test_cross.py", line 168, in _run_test_case
    status = proc.wait()
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\subprocess.py", line 1209, in wait
    return self._wait(timeout=timeout)
  File "C:\Users\iant\miniconda3\envs\bkdev\lib\subprocess.py", line 1506, in _wait
    result = _winapi.WaitForSingleObject(self._handle,

++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ Timeout ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```
This is running x86_64 Windows 11 within Parallels on an arm64 mac, so not the most sensible test system!

Anyway, this is all just an aside really and shouldn't affect the merging of this PR. I just wanted these observations recorded somewhere appropriate.

**Comment by mattpap** [comment]
I started issue #13171 to continue discussion regarding naming conventions for tests.

**Pull request merged**